jax/numpy/linalg.py

```diff
@@ -62,6 +63,53 @@ def svd(a, full_matrices=True, compute_uv=True):
   return lax_linalg.svd(a, full_matrices, compute_uv)
 
 
+@_wraps(onp.linalg.matrix_power)
+def matrix_power(a, n):
+  a = _promote_arg_dtypes(np.asarray(a))
+
+  if a.ndim < 2:
+    raise TypeError("%d-dimensional array given. Array must be at least "
+                    "two-dimensional" % a.ndim)
+  if a.shape[-2] != a.shape[-1]:
+    raise TypeError("Last 2 dimensions of the array must be square")
+  try:
+    n = operator.index(n)
+  except TypeError:
+    raise TypeError("exponent must be an integer")
+
+  if a.dtype != object:
+    fmatmul = np.matmul
+  elif a.ndim == 2:
+    fmatmul = np.dot
+  else:
+    raise NotImplementedError(
+        "matrix_power not supported for stacks of object arrays")
+
+  if n == 0:
+    a = np.empty_like(a)
+    a = ops.index_update(a, ..., np.eye(a.shape[-2], dtype=a.dtype))
```

Comment:
If we have that the original `a` has shape like `(i, j, k, k)` (notice that the final two dimension sizes must be equal), then what we are returning here is essentially an `i` by `j` matrix of `k` by `k` identity matrices. In the case of 2D matrix power, this is essentially equivalent to what you mentioned; however, in higher dimensions (where honestly I cannot wrap my head around what matrix power is actually doing, intuitively) this differs.

I tried thinking of ways to get around this, since these couple lines do bring about this question that you asked at first glance, but nothing came to mind. Any thoughts?